src/shared/sol-iio.c

```diff
@@ -0,0 +1,1094 @@
+    /* Create new trigger */
+    i = rand();
+    if (sol_util_write_file(SYSFS_TRIGGER_SYSFS_ADD_TRIGGER, "%d", i) < 0)
+        goto error;
+
+    /* Set device current trigger */
+    len = asprintf(&trigger_name, "sysfstrig%d", i);
+    if (len < 0)
+        goto error;
+    result = set_current_trigger(device, trigger_name);
+
+    if (result)
+        device->trigger_name = trigger_name;
+    else
+        free(trigger_name);
+
+    return result;
+
+error:
+    SOL_WRN("Could not create sysfs trigger.");
+
+    return false;
+}
+
+static bool
+check_trigger(struct sol_iio_device *device)
+{
+    char path[PATH_MAX];
+    char *name = NULL;
+    int len;
+    bool r;
+
+    r = craft_filename_path(path, sizeof(path), CURRENT_TRIGGER_DEVICE_PATH, device->device_id);
+    if (!r) {
+        SOL_WRN("Could not open IIO device%d trigger on sysfs", device->device_id);
+        return false;
+    }
+
+    len = sol_util_read_file(path, "%ms", &name);
+    if (len < 0) {
+        SOL_INF("No current trigger for iio:device%d. Creating a sysfs one.",
+            device->device_id);
+        if (!create_sysfs_trigger(device))
+            return false;
+    } else
+        device->trigger_name = name;
+
+    return check_manual_triggering(device);
+}
+
+static void
+set_buffer_size(struct sol_iio_device *device, int buffer_size)
+{
+    char path[PATH_MAX];
+
+    if (craft_filename_path(path, sizeof(path), BUFFER_LENGHT_DEVICE_PATH, device->device_id)) {
+        SOL_WRN("Could not set IIO device buffer size");
+        return;
+    }
+
+    if (sol_util_write_file(path, "%d", buffer_size) < 0) {
+        SOL_WRN("Could not set IIO device buffer size");
+    }
+
+    return;
+}
+
+static bool
+set_buffer_enabled(struct sol_iio_device *device, bool enabled)
+{
+    char path[PATH_MAX];
+    bool r;
+
+    r = craft_filename_path(path, sizeof(path), BUFFER_ENABLE_DEVICE_PATH, device->device_id);
+    if (!r)
+        return false;
+
+    if (sol_util_write_file(path, "%d", enabled) < 0)
+        return false;
+
+    return true;
+}
+
+static size_t
+calc_buffer_size(const struct sol_iio_device *device)
+{
+    struct sol_iio_channel *channel;
+    int i;
+    size_t size = 0;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&device->channels, channel, i) {
+        size += channel->storagebits;
+    }
+
+    return size / 8 + (size % 8 != 0);
+}
+
+static bool
+device_reader_cb(void *data, int fd, unsigned int active_flags)
+{
+    struct sol_iio_device *device = data;
+    bool result = true;
+
+    if (result && active_flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL)) {
+        SOL_WRN("Unexpected reading");
+        result = false;
+    }
+
+    while (result) {
+        ssize_t ret;
+
+        ret = read(fd, device->buffer, device->buffer_size);
+        if (ret < 0) {
+            if (errno == EINTR)
+                continue;
+            else if (errno == EAGAIN || errno == EWOULDBLOCK)
+                break;
+            else {
```

Comment:
what about that idea to export the function from `sol-util-linux.c` that does this inner loop to read data?

Also, here you should retry if `EAGAIN || EWOULDBLOCK` (actually they are the same), with a upper bound. You do so because more data may be available right before you exit... then it's not worth to go back to the loop, instead try again couple of times.

BUT, AFAIU you're going to change this to use `mmap()`, in such case you just use the `device_reader_cb()` to be notified about more data, the read would act directly on the mmape'd memory -- no loops at all.